app/vmalert/datasource/datasource.go

```diff
@@ -18,6 +18,7 @@ type Querier interface {
 	// It returns list of Metric in response and error if any.
 	// QueryRange should stop once ctx is cancelled.
 	QueryRange(ctx context.Context, query string, from, to time.Time) ([]Metric, error)
+	Health() bool
```

Comment:
I suggest to consider creating a separate package or just a file with interface `Healther` or `Pinger` with only one method.
Other entities then can implement this interface or not. For example, datasource could implement it while remotewrite - will not.
This file/package should contain a logic that will help to track and cache the Health status of one or more entities that implement the interface. This approach would help to periodically check health of all entities in background and provide the cached values when requested via Web UI.

Reply:
but if i separate package, datasource and notifier interface could refer health pkg, while health pkg should check is  healthy periodically. it could lead to cycle refering.
my another solution is datasource and notifier register IsHealth(function) to health pkg, but if notifier reload, it may lead to mistake.
any other better solution?